lite/verifying/provider.go

```diff
@@ -0,0 +1,440 @@
+type Provider struct {
+	chainID     string
+	logger      log.Logger
+	trustPeriod time.Duration // e.g. the unbonding period, or something smaller.
+	now         nowFn
+	height      int64
+
+	// Already validated, stored locally
+	trusted lite.PersistentProvider
+
+	// New info, like a node rpc, or other import method.
+	source lite.Provider
+
+	// pending map to synchronize concurrent verification requests
+	mtx                  sync.Mutex
+	pendingVerifications map[int64]chan struct{}
+}
+
+// makeProvider returns a new verifying Provider. It uses the
+// trusted Provider to store validated data and the source Provider to
+// obtain missing data (e.g. FullCommits).
+//
+// The trusted Provider should be a DBProvider.
+// The source Provider should be a client.HTTPProvider.
+// NOTE: The external facing constructor is called NewVerifyingProivider.
+func makeProvider(chainID string, trustPeriod time.Duration, trusted lite.PersistentProvider, source lite.Provider, logger log.Logger) *Provider {
+	if trustPeriod == 0 {
+		panic("VerifyingProvider must have non-zero trust period")
+	}
+	logger = logger.With("module", loggerPath)
+	trusted.SetLogger(logger)
+	source.SetLogger(logger)
+	return &Provider{
+		logger:               logger,
+		chainID:              chainID,
+		trustPeriod:          trustPeriod,
+		trusted:              trusted,
+		source:               source,
+		pendingVerifications: make(map[int64]chan struct{}, sizeOfPendingMap),
+	}
+}
+
+// Implements Verifier.
+func (vp *Provider) Verify(signedHeader types.SignedHeader) error {
+
+	// We can't verify commits for a different chain.
+	if signedHeader.ChainID != vp.chainID {
+		return cmn.NewError("BaseVerifier chainID is %v, cannot verify chainID %v",
+			vp.chainID, signedHeader.ChainID)
+	}
+
+	valSet, err := vp.ValidatorSet(signedHeader.ChainID, signedHeader.Height)
+
+	if err != nil {
+		return err
+	}
+
+	// We can't verify commits older than bv.height.
+	if signedHeader.Height < vp.height {
+		return cmn.NewError("BaseVerifier height is %v, cannot verify height %v",
+			vp.height, signedHeader.Height)
+	}
+
+	// We can't verify with the wrong validator set.
+	if !bytes.Equal(signedHeader.ValidatorsHash, valSet.Hash()) {
+		return lerr.ErrUnexpectedValidators(signedHeader.ValidatorsHash, valSet.Hash())
+	}
+
+	// Do basic sanity checks.
+	err = signedHeader.ValidateBasic(vp.chainID)
+	if err != nil {
+		return cmn.ErrorWrap(err, "in verify")
+	}
+
+	// Check commit signatures.
+	err = valSet.VerifyCommit(
+		vp.chainID, signedHeader.Commit.BlockID,
+		signedHeader.Height, signedHeader.Commit)
+	if err != nil {
+		return cmn.ErrorWrap(err, "in verify")
+	}
+
+	return nil
+}
+
+func (vp *Provider) SetLogger(logger log.Logger) {}
+
+func (vp *Provider) ChainID() string {
+	return vp.chainID
+}
+
+// Implements UpdatingProvider
+//
+// On success, it will store the full commit (SignedHeader + Validators) in vp.trusted
+// NOTE: For concurreent usage, use concurrentProvider
```

Comment:
```suggestion
// NOTE: For concurrent usage, use concurrentProvider
```
Also, it is confusing to refer to a private struct in a public comment of a public method.